sycl/source/detail/graph_impl.hpp

```diff
@@ -878,18 +878,12 @@ class graph_impl : public std::enable_shared_from_this<graph_impl> {
   /// Add a queue to the set of queues which are currently recording to this
   /// graph.
   /// @param RecordingQueue Queue to add to set.
-  void
-  addQueue(const std::shared_ptr<sycl::detail::queue_impl> &RecordingQueue) {
-    MRecordingQueues.insert(RecordingQueue);
-  }
+  void addQueue(sycl::detail::queue_impl &RecordingQueue);
```

Comment:
I had to move this and the ones below into the `graph_impl.cpp` because `queue_impl` is incomplete at this point. `queue_impl.hpp`/`graph_impl.hpp` include each other (which is totally wrong but not the subject of this PR) and untangling that would be a task that I'm not willing to take on, at least not now. 

Another alternative would be to make these templates:

```
template <typename QueueImplTy = queue_impl>
void foo(QueueImplTy &Q)
```
to make `weak_from_this()` template-type-dependent and delay the requirement for a complete type until `foo` is instantiated (vs defined as of now).

Reply:
> I had to move this and the ones below into the graph_impl.cpp because queue_impl is incomplete at this point. queue_impl.hpp/graph_impl.hpp include each other (which is totally wrong but not the subject of this PR) and untangling that would be a task that I'm not willing to take on, at least not now.

Wasn't aware of the bidirectional include of `queue_impl.hpp` and `graph_impl.hpp`, thanks for pointing that out. We've been wanting to split graph_impl hpp/cpp up into smaller files as it's gotten quite large, so made a note on our internal ticket for that to remove this bidirectional include.

I think moving these functions into the cpp file is fine though.